### ISSUE TITLE
Fontmake assumes there is no mark feature in UFO’s features.fea

### ISSUE THREAD

**Issue by MrBrezina** [googlei18n/fontmake#319]
When fontmaking TTF from UFO instances, I get the following report. I think it is because fontmake simply adds mark feature to the other features and does not check if mark feature is already there.

```
  File "/usr/local/bin/fontmake", line 11, in <module>
    load_entry_point('fontmake==1.2.5', 'console_scripts', 'fontmake')()
  File "/usr/local/lib/python2.7/site-packages/fontmake/__main__.py", line 195, in main
    ufo_paths, is_instance=args.pop('masters_as_instances'), **args)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 481, in run_from_ufos
    **kwargs)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 193, in build_ttfs
    self.save_otfs(ufos, ttf=True, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
    return func(*args, **kwds)
  File "/usr/local/lib/python2.7/site-packages/fontmake/font_project.py", line 293, in save_otfs
    font = compileTTF(ufo, convertCubics=False, **compiler_options)
  File "/usr/local/lib/python2.7/site-packages/ufo2ft/__init__.py", line 67, in compileTTF
    featureCompiler.compile()
  File "/usr/local/lib/python2.7/site-packages/ufo2ft/featureCompiler.py", line 40, in compile
    self.setupFile_featureTables()
  File "/usr/local/lib/python2.7/site-packages/ufo2ft/featureCompiler.py", line 141, in setupFile_featureTables
    filename=feapath)
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/builder.py", line 25, in addOpenTypeFeaturesFromString
    addOpenTypeFeatures(font, featurefile)
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/builder.py", line 16, in addOpenTypeFeatures
    builder.build()
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/builder.py", line 78, in build
    self.parseTree = Parser(self.file, self.glyphMap).parse()
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/parser.py", line 53, in parse
    statements.append(self.parse_markClass_())
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/parser.py", line 507, in parse_markClass_
    markClass.addDefinition(mcdef)
  File "/usr/local/lib/python2.7/site-packages/fontTools/feaLib/ast.py", line 336, in addDefinition
    definition.location)
fontTools.feaLib.error.FeatureLibError: /Users/davidbrezina/Typefaces (retail)/Avory/production/Avory/instance_ufo/Avory-Regular.ufo/features.fea:5464:1: Glyph apostrophemodifier already defined at /Users/davidbrezina/Typefaces (retail)/Avory/production/Avory/instance_ufo/Avory-Regular.ufo/features.fea:3934:8
```

**Comment by anthrotype** [comment]
The problem is in ufo2ft MarkFeatureWriter, which writes not only mark but also mkmk features, based on the presence of anchors in the UFO, and also based on whether the features are not already there in the features.fea or any included files.

I think you have generated your own mark feature with other tools, but did not also add a mkmk feature, which is fine. Now, the MarkFeatureWriter sees you added a mark feature so it doesn't write that; however it sees some of the mark glyphs have mark-to-mark anchors and thus automatically generates a mkmk feature; but the mark classes ufo2ft creates to make its mkmk feature enter in conflict with the existing mark classes you have already defined for your mark feature, hence the feaLib error above.

Basically, the way things work at the moment, either you don't define neither mark nor mkmk manually and let both be generated by the toolchain, or you define both of them yourself. Only having a mark feature, but also some glyphs containing e.g. both "_top" and "top", will make so that an mkmk feature will be created, which will likely clash with the existing mark classes,  (they usually overlap and are shared by the two features).

I can look at this early next week. One workaround admittedly hackish you have, would be to add an empty feature mkmk {}; statement in your features so that ufo2ft will not attempt to write one (as you remember, feaLib ignores those, but makeotf will crash with syntax error).

**Comment by anthrotype** [comment]
Actually the problem arises also if one does not have any mark-to-mark anchors in the font; it is enough to just have one mark (or mkmk) feature already defined; the markFeatureWriter will then attempt to create the other missing feature; but the problem is it will unconditionally write its own conflicting markClass definitions, even if at the end the feature block itself turns out to be empty and therefore skipped altogether.

I think it would be better if we skip the whole markFeatureWriter if any of mark or mkmk feature are already defined, rather than treating them separately.

**Comment by MrBrezina** [comment]
Ha, funnily, I used ufo2ft to generate mark and mkmk so I could use it for makeotf, but it correctly did not write any mkmk because there were no anchors for that. Bit silly that it does not consider its own output as “complete”. Adding a few anchors to have some mkmk fixed it. Thanks.

**Comment by khaledhosny** [comment]
I think a more robust fix is for the feature writer to check the existing mark classes and make sure the new ones are unique. `ufo2ft` already parses the features using `feaLib` so it is shouldn’t add too much overhead to do such checks.

**Comment by anthrotype** [comment]
> I think a more robust fix is for the feature writer to check the existing mark classes and make sure the new ones are unique.

You're right, but we can do that later. The current functionality of trying to write mark and mkmk independently from each other is broken anyway, so we may well just accept the fact that the current MarkFeatureWriter can either write both mark+mkmk or none if any of the two is already present.
